File: app/src/main/java/com/example/development/sakaiclient20/networking/deserializers/AnnouncementDeserializer.java
package com.example.development.sakaiclient20.networking.deserializers;

import com.example.development.sakaiclient20.persistence.entities.Announcement;
import com.example.development.sakaiclient20.persistence.entities.Attachment;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

public class AnnouncementDeserializer implements JsonDeserializer<Announcement> {


    @Override
    public Announcement deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {

        JsonObject jsonObject = json.getAsJsonObject();
        String id = jsonObject.get("announcementId").getAsString();
        Announcement announcement = new Announcement(id);

        announcement.createdOn = jsonObject.get("createdOn").getAsLong();

        JsonElement createdByElement = jsonObject.get("createdByDisplayName");
        announcement.createdBy = createdByElement.isJsonNull() ?
                "" : createdByElement.getAsString();

        announcement.body = jsonObject.get("body").getAsString();
        announcement.title = jsonObject.get("title").getAsString();
        announcement.siteId = jsonObject.get("siteId").getAsString();

        JsonArray attachments = jsonObject.get("attachments").getAsJsonArray();
        for(int i = 0; i < attachments.size(); i++) {
            JsonObject attachmentObject = attachments.get(i).getAsJsonObject();
            Attachment attachment = context.deserialize(attachmentObject, Attachment.class);
            attachment.assignmentId = announcement.announcementId;
            announcement.attachments.add(attachment);
        }

        return announcement;

    }
}

File: app/src/main/java/com/sakaimobile/development/sakaiclient20/networking/deserializers/AnnouncementDeserializer.kt
package com.sakaimobile.development.sakaiclient20.networking.deserializers

import com.google.gson.JsonArray
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.sakaimobile.development.sakaiclient20.persistence.entities.Announcement
import com.sakaimobile.development.sakaiclient20.persistence.entities.Attachment

import java.lang.reflect.Type

class AnnouncementDeserializer : JsonDeserializer<Announcement> {

    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Announcement {

        val jsonObject = json.asJsonObject
        val id = jsonObject.getStringMember("announcementId")!!
        val announcement = Announcement(id)

        announcement.createdOn = jsonObject.get("createdOn").asLong
        announcement.createdBy = jsonObject.getStringMember("createdByDisplayName")

        announcement.body = jsonObject.getStringMember("body")
        announcement.title = jsonObject.getStringMember("title")
        announcement.siteId = jsonObject.getStringMember("siteId")

        announcement.attachments = jsonObject.get("attachments").asJsonArray.map { attachmentObject ->
            val attachment = context.deserialize<Attachment>(attachmentObject, Attachment::class.java)
            attachment.announcementId = announcement.announcementId
            attachment
        }

        return announcement

    }
}
